core/lib/zksync_core/src/vm_runner/storage.rs

```diff
@@ -0,0 +1,456 @@
+    /// Creates a new VM runner storage using provided Postgres pool and RocksDB path.
+    pub async fn new(
+        pool: ConnectionPool<Core>,
+        rocksdb_path: String,
+        loader: L,
+        enum_index_migration_chunk_size: usize, // TODO: Remove
+        chain_id: L2ChainId,
+        max_batches_to_load: u32,
+    ) -> anyhow::Result<(Self, StorageSyncTask<L>)> {
+        let mut conn = pool.connection_tagged(L::name()).await?;
+        let l1_batch_params_provider = L1BatchParamsProvider::new(&mut conn)
+            .await
+            .context("Failed initializing L1 batch params provider")?;
+        drop(conn);
+        let rocksdb_cell = Arc::new(OnceCell::new());
+        let state = Arc::new(RwLock::new(State {
+            l1_batch_number: L1BatchNumber(0),
+            storage: BTreeMap::new(),
+        }));
+        let task = StorageSyncTask::new(
+            pool.clone(),
+            max_batches_to_load,
+            chain_id,
+            rocksdb_path,
+            enum_index_migration_chunk_size,
+            loader.clone(),
+            rocksdb_cell.clone(),
+            state.clone(),
+        )
+        .await?;
+        Ok((
+            Self {
+                pool,
+                l1_batch_params_provider,
+                chain_id,
+                max_batches_to_load,
+                state,
+                rocksdb_cell,
+                _marker: PhantomData,
+            },
+            task,
+        ))
+    }
+
+    async fn access_storage_inner(
+        &self,
+        stop_receiver: &watch::Receiver<bool>,
+        l1_batch_number: L1BatchNumber,
+    ) -> anyhow::Result<Option<PgOrRocksdbStorage<'_>>> {
+        if let Some(rocksdb) = self.rocksdb_cell.get() {
+            let state = self.state.read().await;
+            let mut conn = self
+                .pool
+                .connection_tagged(L::name())
+                .await
+                .context("Failed getting a Postgres connection")?;
+            let max_l1_batch = (state.l1_batch_number + self.max_batches_to_load - 1).min(
+                conn.blocks_dal()
+                    .get_sealed_l1_batch_number()
+                    .await?
+                    .unwrap_or_default(),
+            );
+            if l1_batch_number < state.l1_batch_number || l1_batch_number > max_l1_batch {
+                anyhow::bail!(
+                    "Trying to access VM runner storage with L1 batch #{} while only [#{}; #{}] are available",
+                    l1_batch_number,
+                    state.l1_batch_number,
+                    state.l1_batch_number + self.max_batches_to_load - 1
+                );
+            }
+            let rocksdb_builder = RocksdbStorageBuilder::from_rocksdb(rocksdb.clone());
+            let rocksdb = rocksdb_builder
+                .synchronize(&mut conn, stop_receiver, Some(state.l1_batch_number))
+                .await
+                .context("Failed to catch up RocksDB storage to Postgres")?;
```

Comment:
RocksDB synchronization ideally shouldn't block access_storage. IIUC we can read and write to RocksDB in parallel.

Reply:
This isn't a "real" synchronization. Currently we have `RocksdbStorage` structured in a way that the only way to get an instance is through running `synchronize` on a builder. I will see if I can refactor this in a way that makes more sense though.

Reply:
I have improved the invariant by including `RocksdbStorage` in the state so it should be better now